### ISSUE TITLE
FixedBucketSampler may return batch with varied length (setting ratio=0)

### ISSUE THREAD

**Issue by khui** [dmlc/gluon-nlp#683]
When using `FixedBucketSampler`, it seems if a bucket includes fewer than `batch_size` data points, the sampler may return a batch with less data (than the configured `batch_size`).  This might be an unexpected behavior for the users since the `ratio=0` and `batch_size=N` are configured to get exactly `N` data points per batch. 

I observed this when using the `FixedBucketSampler` together with the `DataLoader` as in the following, therein making a guess about the reason at this [line](https://github.com/dmlc/gluon-nlp/blob/1821ff95cc12b7ede4928485e9ebfa9ba79cc80b/src/gluonnlp/data/sampler.py#L415). Please correct me if there exist configuration to get batches including the same amount of data.

To repeat the errors, one may simply configure a large enough `num_buckets` (e.g., 20).

```
batchify_fn = mxex.gluonnlp.data.batchify.Dictionary({constants.QUESTION:
                                Pad(axis=0,
                                    pad_val=vocab[constants.PADDING_TOKEN],
                                    ret_length=False,
                                    dtype='int32'),
                            constants.ANSWERS:
                                Pad(axis=0,
                                    pad_val=vocab[constants.PADDING_TOKEN],
                                    ret_length=False,
                                    dtype='int32'),
                            constants.CONTEXTS:
                                Pad(axis=0,
                                    pad_val=vocab[constants.PADDING_TOKEN],
                                    ret_length=False,
                                    dtype='int32')})

batch_sampler = gluonnlp.data.FixedBucketSampler(lengths=txt_lengths,
                                                         batch_size=batch_size,
                                                         num_buckets=num_buckets_in_sampler,
                                                         shuffle=True)
data_loader = gluon.data.DataLoader(encoded_seqs,
                                        batch_sampler=batch_sampler,
                                        batchify_fn=batchify_fn,
                                        num_workers=num_workers)
```

**Comment by szha** [comment]
When ratio is 0, there's still the case where `num_samples % batch_size != 0`. In this case, we'd have the choice of 1) drop the batch, 2) pad samples 3) leave as is. There likely are scenarios that each of these options are good for. @sxjscience @szhengac how do you recommend we handle these cases?

**Comment by khui** [comment]
@szha Thanks for the comments. 

I may not find the `last_batch` parameter for the sampler since the `last_batch` is disabled in `data_loader` if a `batch_sampler` is used.

**Comment by sxjscience** [comment]
The current design is that the data are loaded without any padding. The user may pad the data/drop the batch after it’s loaded. I agree it should be added to the document.

**Comment by khui** [comment]
@sxjscience
 As far as I understand, the padding is orthogonal right? This is the sampler, meanwhile the padding is done with the function in batchify `batchify_fn `.

**Comment by sxjscience** [comment]
@khui Thanks for the question.

1) Yes, I mean the current implementation will not pad the incomplete batches. There are multiple ways to handle those incomplete batches, e.g., padding with zeros, padding by repeating the last few samples, padding by feeding in samples in a round-robin order, or just drop the batch.

We chose not to handle these cases and just implemented the very basic strategy (no padding) because it simplifies the API and can satisfy the requirement of most users. Now we can think if we could add an additional API that deals with these scenarios.

2) I haven't fully understood the second question. Would you provide an example about how TF handles that?

**Comment by khui** [comment]
@sxjscience 
Thanks for the answers. 
As for 1, I am not sure why people will pad zeros to complete the batch. Could you provide examples? The sampler basically randomlly sample (random sample with replacement) the data points to generate batch, thereby, one might assume there wont be not enough data. Namely, even when I only have 10 data point, I should still generate batch with 32 data points, by simply repeating some of the data points. 

For 2), please refer to the `repeat()` in the dataset api. 
 https://www.tensorflow.org/api_docs/python/tf/data/Dataset#repeat.
The use of repeat is something like that:
```
dataset = tf.data.Dataset.from_generator(...)
dataset=dataset.apply(lambda X, _:(transformation_func(X), y))
dataset=dataset.shuffle(buffer_size=10000)
dataset=dataset.repeat() 
iterator = dataset.make_one_shot_iterator()
nxt_X, nxt_y = iterator.get_next()

....
for batch_id in range(num_batch_to_run):
    loss_np = sess.run(loss, feed_dict={X: nxt_X, y: nxt_y})
```

**Comment by sxjscience** [comment]
Sometimes the data/label are all padded to zero to mimic a masking behavior, although that’s not necessary in our case. I think adding a padding by samples/drop/no_padding option would be feasible in our case.

BTW,  would the performance be downgraded if the batchsize are not always the same? Since Gluon support dynamic shapes, I assumed that we would no longer require to pad the data.

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by khui** [comment]
I am not sure about the influcences of the variant batch size on the performance. My point is that the behavior should align with the expectation and the convention as well. Especially notice that, the `batch_size` (together with the `learning rate`) is among the important hyperparameters to govern the learning process.  Up to the distribution of the data, however, it is possible that significant amount of batches are with different size then the configured one (use big bucket number and big batch size).

More importantly, for those bucket that includes not enough samples for a batch, there exist no sampling mechanism when the bucket is chosen, this could be **harmful** to the performance. Thereby, some measurements beyond documentation, e.g., merging those incompleted buckets make more sense, to 1) get enough sampling per batch; 2) actually do the sampling and introduce randomness. 


You are right that the FixedBucketSampler actually supports the varied batch sizes. And better documentation could at lesat remind the users to pay attention to the behaviors of the samplers.

**Comment by Arhamna** [comment]
I am facing a similar issue:

SET UP:
-------------------

train_batch_sampler = nlp.data.FixedBucketSampler(lengths=data_train_lengths,
                                                      batch_size=500,
                                                      num_buckets=1,
                                                      ratio=0.0,
                                                      shuffle=True,
                                                      use_average_length=True,
                                                      num_shards=1,
                                                      bucket_scheme=bucket_scheme)

Where,

len(data_train_lengths) is 3560
bucket_scheme is obtained using "nlp.data.ConstWidthBucket()"


ISSUE:
-------------------

 With the above settings, the number of batches should be 3530/500 = 7.

I, however, see the total number of batches (as returned by the data loader) to be 118!

Additionally, train_batch_sampler.stats() gives a batch size of 30! Not sure why.

> train_batch_sampler.stats():
  sample_num=3530, batch_num=118
  key=[(31, 32)]
  cnt=[3530]
  batch_size=[30]

> train_batch_sampler._batch_size  
500

**Comment by szhengac** [comment]
@Arhamna. When `use_average_length=True`, `batch_size` means the approximate number of tokens in a mini-batch.

**Comment by Arhamna** [comment]
I see. Yes setting use_average_length=False gives the desired batch size. Thank you so much!
